aequilibrae/paths/graph.py

```diff
@@ -103,25 +109,206 @@ def prepare_graph(self, centroids: np.ndarray) -> None:
 
-        all_titles = list(self.network.columns)
-
-        not_pos = self.network.loc[self.network.direction != 1, :]
-        not_negs = self.network.loc[self.network.direction != -1, :]
+        self.__build_compressed_graph()
+        self.compact_num_links = self.compact_graph.shape[0]
+
+    def __build_compressed_graph(self):
+        # Build link index
+        link_idx = np.empty(self.network.link_id.max() + 1).astype(np.int)
+        link_idx[self.network.link_id] = np.arange(self.network.shape[0])
+
+        nodes = np.hstack([self.network.a_node.values, self.network.b_node.values])
+        links = np.hstack([self.network.link_id.values, self.network.link_id.values])
+
+        idx = np.argsort(nodes)
+        all_nodes = nodes[idx]
+        all_links = links[idx]
+        links_index = np.empty(all_nodes.max() + 1, np.int64)
+        links_index.fill(-1)
+        nlist = np.arange(all_nodes.max())
+
+        y, x, _ = np.intersect1d(all_nodes, nlist, assume_unique=False, return_indices=True)
+        links_index[y] = x[:]
+        links_index[-1] = all_links.shape[0]
+        for i in range(all_nodes.max(), 1, -1):
+            links_index[i - 1] = links_index[i] if links_index[i - 1] == -1 else links_index[i - 1]
+
+        nodes = np.hstack([self.network.a_node.values, self.network.b_node.values])
+        counts = np.bincount(nodes)
+        counts[self.centroids] = 999
+
+        truth = (counts == 2).astype(np.int)
+        link_edge = truth[self.network.a_node.values] + truth[self.network.b_node.values]
+        link_edge = self.network.link_id.values[link_edge == 1]
+
+        simplified_links = np.empty(self.network.link_id.max() + 1)
+        simplified_links.fill(-1)
+        simplified_links = simplified_links.astype(np.int)
+        simplified_directions = np.zeros(self.network.link_id.max() + 1, np.int)
+
+        compressed_dir = np.zeros(self.network.link_id.max() + 1, np.int)
+        compressed_a_node = np.zeros(self.network.link_id.max() + 1, np.int)
+        compressed_b_node = np.zeros(self.network.link_id.max() + 1, np.int)
+
+        slink = 0
+        major_nodes = {}
+        tot = 0
+        tot_graph_add = 0
+        for pre_link in link_edge:
+            if simplified_links[pre_link] >= 0:
+                continue
+            ab_dir = 1
+            ba_dir = 1
+            lidx = link_idx[pre_link]
+            a_node = self.network.a_node.values[lidx]
+            b_node = self.network.b_node.values[lidx]
+            drc = self.network.direction.values[lidx]
+            n = a_node if counts[a_node] == 2 else b_node
+            first_node = b_node if counts[a_node] == 2 else a_node
+
+            ab_dir = 0 if (first_node == a_node and drc < 0) or (first_node == b_node and drc > 0) else ab_dir
+            ba_dir = 0 if (first_node == a_node and drc > 0) or (first_node == b_node and drc < 0) else ba_dir
+
+            while counts[n] == 2:
+                if simplified_links[pre_link] >= 0:
```

Comment:
I like assert here, i.e. assert `simplified_links[pre_link] == -1, "your message"`. tomato, tomato.

Reply:
This was actually a debug thing when I was implementing the code. removed it